config.go

```diff
@@ -0,0 +1,186 @@
+		if _, err := url.Parse(svr); err == nil {
+			c.Servers = append(c.Servers, svr)
+		}
+	}
+
+	return c
+}
+
+// MarshalJSON encodes a Config value into a JSON format byte slice.
+func (c *Config) MarshalJSON() ([]byte, error) {
+	m := map[string]interface{}{}
+	if c.DialTimeout != 0 {
+		m["dial_timeout"] = c.DialTimeout.String()
+	}
+
+	m["discover"] = c.Discover
+	if c.Timeout != 0 {
+		m["timeout"] = c.Timeout.String()
+	}
+
+	if c.WatchInterval != 0 {
+		m["watch_interval"] = c.WatchInterval.String()
+	}
+
+	if len(c.Servers) > 0 {
+		m["servers"] = c.Servers
+	}
+
+	return json.Marshal(m)
+}
+
+// UnmarshalJSON decodes a JSON format byte slice into the Config value.
+func (c *Config) UnmarshalJSON(b []byte) error {
+	m := map[string]interface{}{}
+	if err := json.Unmarshal(b, &m); err != nil {
+		return errors.Wrap(err, "unable to unmarshal JSON data")
+	}
+
+	if v, ok := m["dial_timeout"].(string); ok {
+		d, err := time.ParseDuration(v)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse dial timeout")
+		}
+
+		if d < 0 || d > time.Minute {
+			return errors.New("invalid dial timeout value")
+		}
+
+		c.DialTimeout = d
+	}
+
+	if v, ok := m["discover"].(bool); ok {
+		c.Discover = v
+	}
+
+	if v, ok := m["timeout"].(string); ok {
+		d, err := time.ParseDuration(v)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse timeout")
+		}
+
+		if d < 0 || d > (5*time.Minute) {
+			return errors.New("invalid timeout value")
+		}
+
+		c.Timeout = d
+	}
+
+	if v, ok := m["watch_interval"].(string); ok {
+		d, err := time.ParseDuration(v)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse watch interval")
+		}
+
+		if d < 0 || d > (24*time.Hour) {
+			return errors.New("invalid watch interval value")
+		}
+
+		c.WatchInterval = d
+	}
+
+	if v, ok := m["servers"].([]interface{}); ok {
+		for _, vv := range v {
+			if svr, ok := vv.(string); ok {
+				if _, err := url.Parse(svr); err == nil {
+					c.Servers = append(c.Servers, svr)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// WithDialTimeout sets a new dial timeout and returns a pointer to the updated
+// configuration value.
+func (c *Config) WithDialTimeout(t time.Duration) *Config {
+	if t >= 0 && t < time.Minute {
+		c.DialTimeout = t
+	}
+
+	return c
+}
+
+// WithDiscover sets a new discover value and returns a pointer to the updated
+// configuration value.
+func (c *Config) WithDiscover(d bool) *Config {
+	c.Discover = d
+	return c
+}
+
+// WithTimeout sets a new timeout duration and returns a pointer to the updated
+// configuration value.
+func (c *Config) WithTimeout(t time.Duration) *Config {
+	if t >= 0 && t < (5*time.Minute) {
+		c.Timeout = t
+	}
+
+	return c
+}
+
+// WithServers assigns a new list of servers and returns a pointer to the
+// updated configuration value.
+func (c *Config) WithServers(servers ...string) *Config {
+	c.Servers = []string{}
+	for _, svr := range servers {
+		if _, err := url.Parse(svr); err == nil {
```

Comment:
If they don't parse, do we care about any error messages?  We do this in UnmarshalJSON above too.

Reply:
Maybe, but also maybe not. These With* functions would be used when initially building a new Config value, and are intended to be chained together. So, all of them follow a pattern of just not assigning the value passed if it does not pass validation. If we were to having them also return an error, that would prevent being able to chain them together. Also, it is important to keep in mind, that these methods are just used to build a configuration value that is then used to initialize a SnowthClient value, not actually setup the client itself. When the SnowthClient is initialized is when the checks happen that do more than just validate if the address formats are valid. Snowth.NewClient() checks that at least one of the provided servers can be connected, and fails with an error if no servers can be connected. So, if no valid server addresses are provided when setting up a Config value, an error will occur when an attempt is made to use it when initializing a SnowthClient value.